rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -640,14 +646,34 @@ impl IvfSubIndex for HNSW {
             )
         };
 
-        if params.ef < k {
-            return Err(Error::Index {
-                message: "ef must be greater than or equal to k".to_string(),
-                location: location!(),
-            });
-        }
-
-        let results = self.search_basic(query.clone(), k, params.ef, bitmap, storage)?;
+        let keep_count = bitmap.as_ref().map(|b| b.len()).unwrap_or_default() as usize;
+        let results = if keep_count < self.len() * 80 / 100 {
+            log::info!("too many rows filtered, using flat search");
+            let bitmap = bitmap.unwrap();
+            let node_ids = storage
+                .row_ids()
+                .enumerate()
+                .filter_map(|(node_id, row_id)| {
+                    bitmap.contains(*row_id as u32).then_some(node_id as u32)
+                });
+            let dist_calc = storage.dist_calculator(query);
+            let mut heap = BinaryHeap::<Reverse<OrderedNode>>::with_capacity(k);
+            for node_id in node_ids {
+                let dist = dist_calc.distance(node_id).into();
```

Comment:
Nit: calculate distance and sort in two stages might be faster?

Reply:
it will if there are not many rows, we can figure out a good threshold here. will do more benchmark here to get it.